### ISSUE TITLE
Socket connection start up time

### ISSUE THREAD

**Issue by victorjkhalaf** [thumbor/thumbor#1373]
When setting up Thumbor 7 through a socket connection (e.g.: 
`thumbor --fd /tmp/thumbor.sock --conf=/tmp/thumbor.conf &`), we noticed that the start-up time can be on the order of 4-6 seconds. We are running Thumbor on a Lambda running a Python runtime with a containerized service.
We run the above shell script in the Docker entrypoint and then just make sure the sock file is always active between calls.
Lambdas have a cold start time that varies between projects, usually 1-2 seconds max in our cases, but with Thumbor socket we've found the start time to be about 6 seconds, so we had to add a `sleep 6` command to the entrypoint shell script before the runtime can receive Lambda events. This is far too long and poses a big problem as we provision for concurrency during higher traffic loads.
We had to use the socket connection approach because the http Tornado route (we ported AWS' Serverless Image Handler that ran the old Python 2 Thumbor to Python 3 and Thumbor 7) would fail to connect.
Would it be possible to explain a bit how the sock file is initialized and if there is any way to speed up its initialization?

As far as what we are loading from the conf file, we have the following loader:
tc_aws.loaders.s3_loader

For filters we have our own plugin and the following
thumbor.filters.quality
thumbor.filters.curve
thumbor.filters.format

Optimizers:
thumbor_plugins.optimizers.pngquant

For storage and result storage we use tc_redis.

Thanks for any insight!

**Comment by heynemann** [comment]
Can you explain exactly what you mean by not connecting when using ports instead of sock file?

As for the sock file taking too long, I'm not sure on what's going on...

**Comment by victorjkhalaf** [comment]
I apologize for the late response, we decided to try playing around a little bit more and understand the implementation better. We found a solution that improves the situation a little. So I'm closing this with some context and insight.

You might remember a while ago AWS had a version of their project serverless-image-handler that used Python 2 and Thumbor to integrate Thumbor onto Lambdas. Thumbor was especially great because it allowed custom plugin development to be used with that Lambda package.
Since they switched to Sharp and nodejs (and the ability to have custom plugins was lost in the process), and were deprecating Python 2 images, we decided we'd try to port the handler to Python 3 and were fortunate you had version 7 with Python 3 support started.
What we found was that the original implementation that connected to a local sock file using Tornado and requests_unixsocket wouldn't work. I can't recall exactly why but it just failed to start.
So we tried instead to make Thumbor create the sock connection using the --fd argument.
Then the connection would work. It would take 5 seconds to get the socket connection started however.
We found that Lambda cold start times increased by that much time because we had to implement a sleep command on the Docker entrypoint.sh script so it would be enough time for the socket connection to start up and then allow Lambda requests to go in and be fed to Thumbor.
We weren't sure if there was any better way until we tried to get Thumbor to start on a port instead and use Python Requests library to connect. Perhaps it was my lack of insight at the time but I hadn't tried that approach until now and it worked.
It still takes about 2 seconds to get the server up on cold start, so we have to do a bunch of /healthcheck + sleep calls to ensure we can start sending requests. But it's better than before.
It seems this is just how long connections take to start up and we already got way more out of this little platform update project we took on than we could've expected.

I'd be interested in seeing who else has successfully put Thumbor up on Lambdas and how they implemented it.
For reference, we used a dockerized Thumbor project called Minimal Compact Thumbor (https://github.com/MinimalCompact/thumbor), which was helpful in getting most of the image ready for the new Lambda runtime that allows Docker images. We then took the code for AWS' Serverless Image Handler that ran Thumbor/Py2 (it is no longer available online, we just have a fork of it) and modified it to run on Python 3. We then wrapped it in an AWS SAM project with a Dockerfile and entrypoint.sh files as well as the Lambda RIC for Python. This allowed us to package and deploy our Thumbor with our custom plugin to serverless with minimal fuss.
